src/kernel/pci.c

```diff
@@ -259,6 +309,57 @@ static void pci_probe_device(pci_dev dev)
             break;
         }
     }
+    spin_unlock(&pci_lock);
+}
+
+closure_function(4, 0, void, pci_device_remove_complete,
+                 int, bus, int, slot, int, function, thunk, completion)
+{
+    pci_debug("%s\n", __func__);
+    struct pci_dev dev = {
+        .bus = bound(bus),
+        .slot = bound(slot),
+        .function = bound(function),
+    };
+    spin_lock(&pci_lock);
+    int dev_index = pci_dev_find(&dev);
+    if (dev_index >= 0) {
+        deallocate(devices->h, vector_delete(devices, dev_index), sizeof(struct pci_dev));
+        pci_parse_iomem(&dev, false);
+    }
+    spin_unlock(&pci_lock);
+    apply(bound(completion));
+    closure_finish();
+}
+
+void pci_remove_device(pci_dev dev, thunk completion)
+{
+    pci_debug("%s: %02x:%02x:%x\n", __func__, dev->bus, dev->slot, dev->function);
+    spin_lock(&pci_lock);
+    int dev_index = pci_dev_find(dev);
+    pci_driver driver;
+    void *driver_data;
+    if (dev_index >= 0) {
+        pci_dev d = vector_get(devices, dev_index);
+        thunk pci_completion = closure(devices->h, pci_device_remove_complete,
+                                       d->bus, d->slot, d->function, completion);
+        if (pci_completion != INVALID_ADDRESS)
+            completion = pci_completion;
+        else
+            msg_err("cannot allocate completion closure\n");
+        driver = d->driver;
+        driver_data = d->driver_data;
+        d->driver_data = 0;
+    } else {
+        driver = 0;
+    }
+    spin_unlock(&pci_lock);
+    if (driver) {
+        if (driver->remove && driver_data)
+            apply(driver->remove, driver_data, completion);
+    } else {
+        apply(completion);
```

Comment:
Would it be useful to make completion a status_handler and pass some indication of whether the removal actually happened to acpi_eject?

Reply:
As far as I can see, the only case where the completion is invoked after not having removed the device from the list of probed devices is if pci_remove_device() is called more than once for the same device (perhaps the hypervisor might do some sort of retry if the eject doesn't happen after a timeout), and one call occurs simultaneously with the invocation of the completion from a previous call. As such, this seems a pretty unlikely edge case, I would guess the call to the _EJ0 ACPI method would fail but nothing else would happen except a msg_err().
The other case where `driver` is null is when the kernel doesn'tt bind any driver to a given device; in that case, we simply eject the device right away and nothing should change for acpi_eject.
So I don't see an advantage in adding a status to the completion closure.